src/rt/region/region_rc.h

```diff
@@ -0,0 +1,408 @@
+     **/
+    template<size_t size = 0>
+    static Object* create(Alloc& alloc, const Descriptor* desc)
+    {
+      void* p = alloc.alloc<vsizeof<RegionRc>>();
+      Object* o = Object::register_object(p, RegionRc::desc());
+      auto reg = new (o) RegionRc();
+      reg->use_memory(desc->size);
+
+      if constexpr (size == 0)
+        p = alloc.alloc(desc->size);
+      else
+        p = alloc.alloc<size>();
+      o = Object::register_object(p, desc);
+
+      reg->init_next(o);
+      o->init_iso();
+      o->set_region(reg);
+
+      assert(Object::debug_is_aligned(o));
+      return o;
+    }
+
+    /**
+     * Allocates an object `o` of type `desc` in the region represented by the
+     * Iso object `in`, and adds it to the appropriate ring. Returns a pointer
+     * to `o`.
+     *
+     * The default template parameter `size = 0` is to avoid writing two
+     * definitions which differ only in one line. This overload works because
+     * every object must contain a descriptor, so 0 is not a valid size.
+     **/
+    template<size_t size = 0>
+    static Object* alloc(Alloc& alloc, Object* in, const Descriptor* desc)
+    {
+      assert((size == 0) || (size == desc->size));
+      RegionRc* reg = get(in);
+
+      assert(reg != nullptr);
+
+      void* p = nullptr;
+      if constexpr (size == 0)
+        p = alloc.alloc(desc->size);
+      else
+        p = alloc.alloc<size>();
+
+      auto o = (Object*)Object::register_object(p, desc);
+      assert(Object::debug_is_aligned(o));
+
+      reg->track_object(o, alloc);
+
+      // GC heuristics.
+      reg->use_memory(desc->size);
+      return o;
+    }
+
+    static void incref(Object* o)
+    {
+      RefCount* rc = o->get_ref_count();
+      rc->metadata += 1;
+    }
+
+    static bool decref(Alloc& alloc, Object* o, Object* in)
+    {
+      if (decref_inner(o))
+      {
+        dealloc_object(alloc, o, in);
+        return true;
+      }
+      return false;
+    }
+
+    static uintptr_t debug_get_ref_count(Object* o)
+    {
+      RefCount* rc = o->get_ref_count();
+      return rc->metadata;
+    }
+
+  private:
+    inline static bool decref_inner(Object* o)
+    {
+      RefCount* rc = o->get_ref_count();
+      if (rc->metadata == 1)
+      {
+        return true;
+      }
+      rc->metadata -= 1;
+      return false;
+    }
+
+    static void dealloc_object(Alloc& alloc, Object* o, Object* in)
+    {
+      // We first need to decref -- and potentially deallocate -- any object
+      // pointed to through `o`'s fields.
+      ObjectStack dfs(alloc);
+      ObjectStack sub_regions(alloc);
+      o->trace(dfs);
+      ObjectStack flzr_q(alloc);
+
+      flzr_q.push(o);
+
+      while (!dfs.empty())
+      {
+        Object* p = dfs.pop();
+        switch (p->get_class())
+        {
+          case Object::ISO:
+            sub_regions.push(p);
+            break;
+          case Object::MARKED:
+          case Object::UNMARKED:
+          case Object::SCC_PTR:
```

Comment:
I am not convinced SCC_PTR, RC and Cown should be decreffed immediately.  Also, don't they need a different decref?

Reply:
I think you're right. What about this fd926797dec3db8dd8214fa80f5577f2c147ccf1? Is there something in particular that needs to happen for SCC objects too?